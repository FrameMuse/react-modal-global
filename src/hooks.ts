/*

MIT License

Copyright (c) 2023 Valery Zinchenko

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

*/

import { useContext, useEffect, useState } from "react"

import { modalContext } from "./context"
import { ModalController } from "./ModalController"
import { ModalWindow } from "./ModalWindow"
import { ModalComponent, ModalState } from "./types"

/**
 * Used inside a modal component to access the modal context (`ModalWindow`).
 *
 * Accepts a generic type that is used to infer the props of the modal component.
 * It has 3 overloads:
 * 1. `useModalWindow<typeof ModalClassComponent>()` - infers the props from the class component type.
 * 2. `useModalWindow<typeof ModalFunctionComponent>()` - infers the props from the function component type.
 * 3. `useModalWindow<{ c: 3 }>()` - you can enter props by yourself too.
 */
export function useModalWindow<T>(): ModalWindow<T extends ModalComponent<infer Props> ? Props : T> {
  const context = useContext(modalContext)
  if (!context) throw new Error(`ModalError: ${useModalWindow.name} must be used within a modal context.`)

  // It's safe to case to `any` here because the context is always set to a `ModalWindow`
  // and the arbitrary type `T` is difened by the user.
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return context as ModalWindow<any>
}



const DEFAULT_STATE: ModalState = {
  active: false,
  windows: []
}

/**
 * Provides access to modal state and listens to updates.
 */
export function useModalState(controller: ModalController): ModalState {
  const [modalState, setModalState] = useState(DEFAULT_STATE)

  useEffect(() => {
    return controller.observe(setModalState)
  }, [controller])

  return modalState
}
