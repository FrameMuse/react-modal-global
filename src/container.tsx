import React from "react"
import { classWithModifiers } from "./utils"
import { PopupContext } from "./context"
import { PopupWindow } from "./interfaces"
import { PopupPrivate } from "./controller"

export interface PopupContainerProps {
  className?: string
}
export interface PopupContainerState {
  isActive: boolean
  queue: PopupWindow[]
}

export class PopupContainer extends React.Component<PopupContainerProps, PopupContainerState> {
  state: PopupContainerState = {
    isActive: false,
    queue: []
  }

  constructor(props: any) {
    super(props)
    // Set Popup dispatcher
    PopupPrivate.dispatch = this.setState.bind(this)
  }

  render() {
    const { isActive, queue } = this.state
    const lastPopup = queue[queue.length - 1]
    const { component: PopupWindowComponent, params = {}, close } = lastPopup || {}

    const className = this.props.className || "popup"
    return (
      <div className={classWithModifiers(className, isActive && "active")}>
        <div className={className + "__container"} onClick={close}>
          <div className={className + "__inner"} onClick={event => event.stopPropagation()}>
            <PopupContext.Provider value={lastPopup}>
              {PopupWindowComponent && <PopupWindowComponent {...params} />}
            </PopupContext.Provider>
          </div>
        </div>
      </div>
    )
  }
}
