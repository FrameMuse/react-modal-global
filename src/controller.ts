import { Dispatch, SetStateAction } from "react";
import { PopupContainerState } from "./container";
import { PopupComponent, PopupParams, PopupWindow } from "./interfaces"

type AnyIfEmpty<T extends object> = keyof T extends never ? any : T;

export const PopupPrivate: {
  dispatch: Dispatch<SetStateAction<PopupContainerState>>
} = {
  dispatch: () => { }
}

export class Popup {
  public static open
    <P extends object = {}, AC extends Partial<PopupParams> & P = Partial<PopupParams> & P>
    (component: PopupComponent<P>, ...[params]: (AnyIfEmpty<P> extends object ? [AC] : [AC?])): Promise<void> {
    return new Promise<void>(function (resolve) {
      const popupWindow = { component, params, close }
      Popup.addToQueue(popupWindow)
      function close() {
        resolve()
        Popup.removeFromQueue(popupWindow)
      }
    })
  }
  private static addToQueue(popupWindow: PopupWindow<any>) {
    PopupPrivate.dispatch(state => ({
      isActive: true,
      queue: [...state.queue, popupWindow]
    }))
  }
  private static removeFromQueue(popupWindow: PopupWindow<any>) {
    PopupPrivate.dispatch(state => {
      const queue = state.queue.filter(pw => pw !== popupWindow)
      return {
        isActive: queue.length > 0,
        queue
      }
    })
  }
  public static closeAll() {
    PopupPrivate.dispatch(state => {
      state.queue.forEach(popup => popup.close())
      return {
        isActive: false,
        queue: []
      }
    })
  }
}
