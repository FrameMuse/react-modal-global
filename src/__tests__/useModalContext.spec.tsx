import { render } from "@testing-library/react"
import { act } from "react-dom/test-utils"

import { ModalContainer } from "../container.new"
import { ModalController } from "../controller"
import { useModalContext } from "../hooks"

let Modal: ModalController = new ModalController()
let container = render(<ModalContainer controller={Modal} />).container

beforeEach(() => {
  Modal = new ModalController()
  container = render(<ModalContainer controller={Modal} />).container
})

describe("useModalContext", () => {
  it("should throw error", () => {
    function ComponentWithoutModalContext() {
      useModalContext()
      return null
    }

    const renderCallback = () => render(<ComponentWithoutModalContext />)
    expectToThrow(renderCallback, /useModalContext must be used within a modalContext/)
  })

  it("should return isClosed: false", () => {
    function ComponentWithModalContext() {
      const { isClosed: closed } = useModalContext()

      return <span>{String(closed)}</span>
    }

    act(() => {
      Modal.open(ComponentWithModalContext)
    })

    const closedSpan = container.querySelector("span")?.textContent
    expect(closedSpan).toBe("false")
    expect(container).toMatchSnapshot()
  })

  it("should return isClosed: true", () => {
    function ComponentWithModalContext() {
      const { isClosed: closed } = useModalContext()

      return <span>{String(closed)}</span>
    }

    act(() => {
      Modal.open(ComponentWithModalContext).close()
    })

    expect(Modal.isOpen).toBe(false)
    const closedSpan = container.querySelector("span")?.textContent
    expect(closedSpan).toBe("true")
    expect(container.querySelector(".modal")?.className).toBe("modal")
    expect(container).toMatchSnapshot()
  })
})

/**
 * Helps prevent error logs blowing up as a result of expecting an error to be thrown,
 * when using a library (such as enzyme)
 *
 * @param func Function that you would normally pass to `expect(func).toThrow()`
 */
export const expectToThrow = (func: () => unknown, error?: JestToErrorArg): void => {
  // Even though the error is caught, it still gets printed to the console
  // so we mock that out to avoid the wall of red text.
  const spy = jest.spyOn(console, "error")
  spy.mockImplementation(() => void 0)

  expect(func).toThrow(error)

  spy.mockRestore()
}

type JestToErrorArg = Parameters<jest.Matchers<unknown, () => unknown>["toThrow"]>[0];
