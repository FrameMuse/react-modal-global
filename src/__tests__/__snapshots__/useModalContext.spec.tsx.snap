// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useModalContext should return \`closed\`: false 1`] = `
<div>
  <div
    aria-hidden="false"
    aria-modal="true"
    class="modal modal--active"
  >
    <div
      class="modal__container"
    >
      <span>
        false
      </span>
    </div>
  </div>
</div>
`;

exports[`useModalContext should return \`closed\`: true 1`] = `
<div>
  <div
    aria-hidden="true"
    aria-modal="true"
    class="modal"
  >
    <div
      class="modal__container"
    >
      <span>
        true
      </span>
    </div>
  </div>
</div>
`;
